Framework/PythonInterface/plugins/algorithms/FindMultipleUMatrices.py

```diff
@@ -0,0 +1,380 @@
+        logger.information("H, K, L, d-spac (Ang)")
+        logger.information(f"{np.c_[hkls, np.round(dhkls, 3)]}")
+        logger.information(f"Observed peak d-spacings within limits: {dmin} < d < {dmax}")
+        logger.information(f"{np.round(peaks_filtered.column('DSpacing'), 3)}")
+
+        # loop over every pair of peaks
+        prog_reporter = Progress(self, start=0.0, end=1.0, nreports=len(dhkls))
+        ipairs = list(combinations(range(peaks_filtered.getNumberPeaks()), 2))
+        ubs = []
+        hkl_ers = []
+        for ipair, ipks in enumerate(ipairs):
+            prog_reporter.report()
+            pks = [peaks_filtered.getPeak(ii) for ii in ipks]
+            angle_observed = pks[0].getQLabFrame().angle(pks[1].getQLabFrame())  # between observed peaks
+            if angle_observed > min_angle:
+                # this means don't have to worry about reflections being along same direction
+                # get index of reflections with consistent d-spacing
+                ihkls = [np.flatnonzero(abs(dhkls - pks[ii].getDSpacing()) < dtol) for ii in range(2)]
+                if all(irefl.size > 0 for irefl in ihkls):
+                    # peaks are both consistent with d-spacing of at least one reflection
+                    # loop over every pair of consistent reflections
+                    ihkl_pairs = list(product(*ihkls))
+                    for ihkl_pair in ihkl_pairs:
+                        # generate angles between equivalent hkls for these reflections
+                        hkls1 = ptgr.getEquivalents(hkls[ihkl_pair[0]])
+                        hkls2 = ptgr.getEquivalents(hkls[ihkl_pair[1]])
+                        angles = self.calculate_angles_between_reflections(hkls1, hkls2, unit_cell)
+                        # get a single pair of HKL consistent with observed angle
+                        ihkl1, ihkl2 = np.unravel_index(np.argmin(abs(angles - angle_observed)), angles.shape)
+                        angle_hkl = angles[ihkl1, ihkl2]
+                        if angle_hkl > min_angle and abs(angle_hkl - angle_observed) < angle_tol:
+                            # set hkl of peaks
+                            [pk.setHKL(0, 0, 0) for pk in peaks_filtered]  # reset HKL of all peaks
+                            pks[0].setHKL(*hkls1[ihkl1])
+                            pks[1].setHKL(*hkls2[ihkl2])
+                            try:
+                                self.exec_child_alg("CalculateUMatrix", PeaksWorkspace=peaks_filtered, **self.calc_u_kwargs)
+                            except ValueError:
+                                continue  # skip this pair
+                            # SetUB in all found peaks and calc hkl_ers
+                            this_ub = peaks_filtered.sample().getOrientedLattice().getUB().copy()
+                            self.exec_child_alg("SetUB", Workspace=peaks, UB=this_ub)
+                            nindexed, *_ = self.exec_child_alg(
```

Comment:
This is interesting syntax I haven't seen before! What does `*_` do?

Reply:
Means ignore all the subsequent output instead of having `_` for each output - I think it's relatively new syntax that came in around the same time as the `match case` stuff?

Reply:
Ah excellent, very cool 😎